#Problem 1(a).  Reading Amazon Reviews.

In this problem, we will analyze Amazon reviews to determine what characteristics make them most helpful.

Download the file of Amazon gourmet food reviews from the [Stanford Large Network Dataset Collection](https://snap.stanford.edu/data/web-FineFoods.html).   (Your computer may already have a utility installed that can unzip the archive as a text file; if not, [7-zip](http://www.7-zip.org/) is a useful utility for Windows. You can also use an online utility by doing a web search for: ``open .gz files online``.)

Create a pandas DataFrame object with the following entries for each review:

* Product ID
* Number of people who voted this review helpful
* Total number of people who rated this review
* Rating of the product
* Text of the review

For the second and third of these, the information will be given in the file as ```1/5```, which would correspond to 1 vote for helpful out of 5 people who rated the review.


In [1]:
import numpy as np
import pandas as pd
import os

# function to iterate through foodFile and 'select' each record

def grab_entry(foodFile):
    record = {'Product ID':-1,'Number of people who voted this review helpful':'', 'Total number of people who rated this review':'','Rating of the product':'', 'Text of the review':''}
    line = foodFile.readline()
    
    while line:
        if "product/productId" in line:
            s = line.split(':')
            record['Product ID'] = s[1]
            
        if "review/helpfulness" in line:
            s = line.split(':')
            s = s[1].split('/')
            record['Number of people who voted this review helpful'] = eval(s[0])
            
        if "review/helpfulness" in line:
            s = line.split('/')[2]
            record['Total number of people who rated this review'] = eval(s[-2])
            
        if "review/score" in line:
            s = line.split(':')
            record['Rating of the product'] = eval(s[1])
        
            
        if "review/text" in line:
            s = line.split('review/text: ')
            record["Text of the review"] = str(s[1:])
            
        if line == '\n':
            return record
        line = foodFile.readline()
        
fileLoc = "/Users/gizzidan/Documents/datascience/foodstuff.txt"
data = []
with open(fileLoc) as foodFile:
    fs = os.fstat(foodFile.fileno()).st_size
    num_read = 0
    while not foodFile.tell() == fs:
        data.append(grab_entry(foodFile))
        num_read += 1    
    
food_df = pd.DataFrame(data)


#Problem 1(b).  Analyzing review text.

Add columns to your DataFrame for the length of a review, the number of exclamation points in a review, and the fraction of people who rated a review helpful. You should calculate the fraction who rated a review helpful using the two columns you made in 1a, and a ratio of 1 helpful rating out of 5 total ratings should be entered as 0.2, not the string ```1/5```. If no people voted on whether a problem was helpful, the corresponding entry in your percentage helpful column should be ```NaN```.

In [2]:
length_of_review = [len(row) for row in food_df["Text of the review"]]
food_df["Length of review in characters"] = length_of_review

exclamation_count = [row.count('!') for row in food_df["Text of the review"]]
food_df["Number of exclamation points"] = exclamation_count

for row in food_df:
    fraction = food_df['Number of people who voted this review helpful']/food_df['Total number of people who rated this review']

food_df["Fraction of people who rated this review helpful"] = fraction
food_df


Number of people who voted this review helpful     Product ID  \
0                                                    1   B001E4KFG0\n   
1                                                    0   B00813GRG4\n   
2                                                    1   B000LQOCH0\n   
3                                                    3   B000UA0QIQ\n   
4                                                    0   B006K2ZZ7K\n   
5                                                    0   B006K2ZZ7K\n   
6                                                    0   B006K2ZZ7K\n   
7                                                    0   B006K2ZZ7K\n   
8                                                    1   B000E7L2R4\n   
9                                                    0   B00171APVA\n   
10                                                   1   B0001PB9FE\n   
11                                                   4   B0009XLVG0\n   
12                                                   1   B0009XLVG0\n   
13                                                   2   B001GVISJM\n   
14                                                   4   B001GVISJM\n   
15                                                   4   B001GVISJM\n   
16                                                   0   B001GVISJM\n   
17                                                   0   B001GVISJM\n   
18                                                   0   B001GVISJM\n   
19                                                   0   B001GVISJM\n   
20                                                   0   B001GVISJM\n   
21                                                   0   B001GVISJM\n   
22                                                   0   B001GVISJM\n   
23                                                   0   B001GVISJM\n   
24                                                   0   B001GVISJM\n   
25                                                   0   B001GVISJM\n   
26                                                   0   B001GVISJM\n   
27                                                   0   B001GVISJM\n   
28                                                   0   B00144C10S\n   
29                                                   1   B0001PB9FY\n   
...                                                ...            ...   
568424                                               2   B001FPT1WM\n   
568425                                               2   B001FPT1WM\n   
568426                                               1   B001FPT1WM\n   
568427                                               1   B001FPT1WM\n   
568428                                               1   B001FPT1WM\n   
568429                                               1   B001FPT1WM\n   
568430                                               0   B001FPT1WM\n   
568431                                               0   B003XUL27E\n   
568432                                               0   B003XUL27E\n   
568433                                               0   B003XUL27E\n   
568434                                               0   B003XUL27E\n   
568435                                               0   B003XUL27E\n   
568436                                              10   B000NY4SAG\n   
568437                                               4   B000NY4SAG\n   
568438                                               3   B000NY4SAG\n   
568439                                               0   B000NY4SAG\n   
568440                                               1   B005ZC0RRO\n   
568441                                               0   B000NY8O9M\n   
568442                                               0   B006T7TKZO\n   
568443                                               0   B000H7K114\n   
568444                                               3   B001EO7N10\n   
568445                                               2   B001EO7N10\n   
568446                                               0   B001EO7N1

#Problem 1(c).  Summary statistics.

How many reviews are in the data set?  What is the average length of a review (in characters)?  What is the average rating?  What is the greatest number of exclamation marks used in a single review?  Use the pandas package to answer these questions, then summarize your results in a markdown cell.

In [3]:
print(num_read)  

568454


In [4]:
np.mean(food_df["Length of review in characters"])

442.6563996383173

In [5]:
np.mean(food_df["Rating of the product"])

4.183198640523243

In [6]:
max(food_df["Number of exclamation points"])

84

After performing my analysis above, there are 568,454 number of reviews, the mean length of the reviews is 442.7 characters, the mean rating of the products is 4.2, and the maximum number of exclamation points used in a review is 84.

#Problem 1(d).  Export.

Save your DataFrame as a .csv file suitable for future analysis in R.  Your .csv file should not include the review text column, as the presence of commas and quotation marks will make reading the file difficult.  You should also convert entries from ```NaN``` to the empty string before saving.

In [18]:
food_df.fillna("", inplace=True)
food_df2 = food_df.drop("Text of the review", axis=1)
food_df2.to_csv("food_clean.csv", index=False)